### PR TITLE
bounded_vec: Avoid unnecessary allocation when decoding

### PR DESCRIPTION
**Pull request by mrcnski**
Needed for https://github.com/paritytech/polkadot/pull/6603. See https://github.com/paritytech/polkadot/pull/6603#discussion_r1083461147:

> You want to ensure that you don't even start try decoding/allocating when the length of the vector is more than the
> allowed maximum.

> You first decode length of the vector and then early reject if that is too long.

### CODE REVIEW

File: bounded-collections/src/bounded_vec.rs
Comment (ggwpez) [inline bounded-collections/src/bounded_vec.rs:293]: `BoundedBTreeMap` and `BoundedBTreeSet` also suffer from this.  
Could you please also fix it there?

**Review by bkchr — APPROVED** [review]
% what @ggwpez said.

**Review by KiChjang — APPROVED** [review]
Also probably requires a version bump for substrate to pick up the changes.

**Comment by mrcnski** [comment]
> Also probably requires a version bump for substrate to pick up the changes.

Should I do that in this PR?

Also, I'll try to get to `BoundedBTreeMap` and `BoundedBTreeSet` soon, haven't had a chance yet.

**Review by ggwpez — APPROVED** [review]
>Should I do that in this PR?

You can also do it in another PR, just thought it would fit in here :smile:

**Comment by bkchr** [comment]
As the crate is currently lacking test, I would be happy if this pr could start with this ;) We could ensure that we don't consume more data than the compact length.

**Comment by mrcnski** [comment]
@bkchr There are tests, including for that case -- see `too_big_vec_fail_to_decode` and `too_big_fail_to_decode`. Or do you mean we should check whether unnecessary memory is consumed (allocated)? I wasn't sure how to do that, or if such a test was necessary.

**Comment by mrcnski** [comment]
This PR is already approved, but would like a quick re-approval from someone (to sanity check the last two commits). :)

**Comment by ggwpez** [comment]
>I wasn't sure how to do that, or if such a test was necessary.

Dump idea: Just use `u32` as Bound and `1MiB` as value, then try to decode that. Should fail to allocate if the implementation is wrong.

**Comment by bkchr** [comment]
> I wasn't sure how to do that, or if such a test was necessary.

```
let data = something.encode();
let data_input = &mut &data[..];

Something::decode(data_input).unwrap_err();
assert_eq!(data_input.len(), data.len() - Compact::<u32>(data.len() as u32).compact_len());
```

This should do it. Also some easy tests ensuring that `Vec` etc encodings are the same as the bounded impls.

> There are tests, including for that case -- see `too_big_vec_fail_to_decode` and `too_big_fail_to_decode`

Sorry :see_no_evil:

File: bounded-collections/src/bounded_btree_map.rs
Comment (ggwpez) [inline bounded-collections/src/bounded_btree_map.rs:51]: No idea if rust handles this in the most efficient way, otherwise we should maybe pre-allocate as optimization.

File: bounded-collections/src/bounded_btree_set.rs
Comment (ggwpez) [inline bounded-collections/src/bounded_btree_set.rs:49]: Have not seen these functions yet, but LGTM

**Review by ggwpez — APPROVED** [review]

File: bounded-collections/src/bounded_btree_map.rs
Comment (bkchr) [reply bounded-collections/src/bounded_btree_map.rs:51]: I mean this is now discusable with the bound, but for the normal `BTreeMap` we don't have done this as someone could send you a fake encoded `BTreeMap` that pretends to be `u32::Max` in size and then maybe leading to some OOM.

Comment (mrcnski) [reply bounded-collections/src/bounded_btree_map.rs:51]: Interesting, would have to compile with optimizations and see. At any rate, I don't want to deviate too much from the standard `Decode` for the unbounded types, at least in this PR.

Comment (bkchr) [reply bounded-collections/src/bounded_btree_map.rs:51]: Should be fine to merge as is.

File: bounded-collections/src/bounded_btree_map.rs
Comment (bkchr) [inline bounded-collections/src/bounded_btree_map.rs:32]:
```suggestion
#[cfg_attr(feature = "std", derive(Hash, serde::Serialize, serde::Deserialize), serde(transparent))]
```

File: bounded-collections/src/bounded_btree_set.rs
Comment (bkchr) [inline bounded-collections/src/bounded_btree_set.rs:32]:
```suggestion
#[cfg_attr(feature = "std", derive(Hash, serde::Serialize, serde::Deserialize), serde(transparent))]
```

File: bounded-collections/src/bounded_vec.rs
Comment (bkchr) [inline bounded-collections/src/bounded_vec.rs:43]:
```suggestion
#[cfg_attr(feature = "std", derive(Hash, serde::Serialize, serde::Deserialize), serde(transparent))]
```

**Review by bkchr — APPROVED** [review]
Ty! But let us use the same derives everywhere.

File: bounded-collections/src/bounded_btree_map.rs
Comment (mrcnski) [reply bounded-collections/src/bounded_btree_map.rs:32]: I'm not sure about deriving these; `Deserialize` would need a custom `impl` to add bounds checks, like we do now with `BoundedVec`. And the custom `impl` looks annoying to implement, so I'm thinking we don't add it until there's actually a need for it.

File: bounded-collections/src/bounded_btree_map.rs
Comment (KiChjang) [reply bounded-collections/src/bounded_btree_map.rs:51]: Is this just another way of calling `collect`?

```suggestion
		let inner = (0..len).map(|_| Decode::decode(input)).collect::<Result<Vec<_>, codec::Error>>()?;
```

Comment (mrcnski) [reply bounded-collections/src/bounded_btree_map.rs:51]: Yes, I think it is. http://doc.rust-lang.org/1.66.1/std/result/enum.Result.html#method.from_iter

Currently this impl is the same as the unbounded version, apart from the length check. I think it should be kept the same unless there's a good reason not to.

File: bounded-collections/src/bounded_btree_map.rs
Comment (mrcnski) [reply bounded-collections/src/bounded_btree_map.rs:32]: @bkchr I'm going to go ahead and merge -- these derives can be a follow-up if we really want them.

**Comment by mrcnski** [comment]
@KiChjang Looks like I'm not able to publish `0.1.4`, you're the only listed owner. Possible to add [paritytech/Core devs](https://crates.io/teams/github:paritytech:core-devs) as an owner?

**Comment by KiChjang** [comment]
I've added the core-devs org as the owner for the crate on crates.io.

**Pull request merged**
